File: sdks/java/sdk/src/main/java/gcba/ratings/sdk/Ratings.java
package gcba.ratings.sdk;

import com.goebl.david.Response;
import com.goebl.david.Webb;
import com.goebl.david.WebbException;
import com.google.gson.Gson;
import org.json.JSONObject;

import java.util.HashMap;

/**
 * Created by ritazerrizuela on 7/17/17.
 */

public final class Ratings {
    public Ratings(String api, String app, String platform, String range, String token) {
        validateUrl(api);
        validateKey(app, "app");
        validateKey(platform, "platform");
        validateKey(range, "range");
        validateToken(token);

        this.url = api.trim();
        this.app = app.trim();
        this.platform = platform.trim();
        this.range = range.trim();
        this.token = token.trim();
        this.timeout = 3000;
    }

    private String url;
    private String app;
    private String platform;
    private String range;
    private String token;
    private Integer timeout;
    private HashMap<String, String> user;

    private void validateUrl(String url) {
        if (url == null) throw new IllegalArgumentException("endpoint can't be null");
        if (!url.trim().matches("^(ftp|http|https):\\/\\/[^ \"]+$")) throw new IllegalArgumentException("invalid endpoint");
    }

    private void validateKey(String key, String description) {
        if (key == null) throw new IllegalArgumentException(description + " can't be null");
        if (key.trim().length() != 32) throw new IllegalArgumentException(description + " is not a valid key");
    }

    private void validateToken(String token) {
        if (token == null) throw new IllegalArgumentException("token can't be null");
        if (token.trim().length() < 1) throw new IllegalArgumentException("invalid token");
    }

    private RatingsError validateName(String name) {
        if (name.trim().length() < 3) return RatingsValidationError.NAME_TOO_SHORT;
        if (name.trim().length() > 70) return RatingsValidationError.NAME_TOO_LONG;

        return null;
    }

    private RatingsError validateEmail(String email) {
        if (!email.trim().matches("^[a-zA-Z0-9.!#$%&’*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\\.[a-zA-Z0-9-]+)*$")) return RatingsValidationError.INVALID_EMAIL;
        if (email.trim().length() < 3) return RatingsValidationError.EMAIL_TOO_SHORT;
        if (email.trim().length() > 100) return RatingsValidationError.EMAIL_TOO_LONG;

        return null;
    }

    private RatingsError validateMibaId(String mibaId) {
        if (mibaId.trim().length() != 36) return RatingsValidationError.INVALID_MIBAID;

        return null;
    }

    private RatingsError validateUser(HashMap<String, String> user) {
        String email, mibaId, name;

        name = user.get("name");
        email = user.get("email");
        mibaId = user.get("mibaId");

        if (email == null && mibaId == null) return RatingsValidationError.MISSING_EMAIL_AND_MIBAID;
        if (name != null) return validateName(name);
        if (email != null) return validateEmail(email);

        return validateMibaId(mibaId);
    }

    private RatingsError validateRating(int rating) {
        if (rating < -127 || rating > 127) return RatingsValidationError.INVALID_RATING;

        return null;
    }

    private RatingsError validateDescription(String description) {
        if (description.trim().length() < 3) return RatingsValidationError.DESCRIPTION_TOO_SHORT;
        if (description.trim().length() > 30) return RatingsValidationError.DESCRIPTION_TOO_LONG;

        return null;
    }

    private RatingsError validateComment(String comment) {
        if (comment.trim().length() < 3) return RatingsValidationError.COMMENT_TOO_SHORT;
        if (comment.trim().length() > 1000) return RatingsValidationError.COMMENT_TOO_LONG;

        return null;
    }

    private RatingsRequest getRequest(int rating) {
        RatingsRequest request;

        request = new RatingsRequest();
        request.rating = rating;
        request.range = range;

        request.app.put("key", app);
        request.platform.put("key", platform);

        return request;
    }

    private RatingsResult send(RatingsRequest request) {
        Webb webb;
        Gson gson;
        String json;
        Response<JSONObject> response;
        Response errorResponse;
        RatingsHTTPError error;

        webb = Webb.create();
        gson = new Gson();
        json = gson.toJson(request);

        webb.setBaseUri(url);

        try {
            response = webb
                .post("/ratings")
                .header("Content-Type", "application/json; charset=UTF-8")
                .header("Accept", "application/json")
                .header("Authorization", "Bearer " + token)
                .body(json)
                .connectTimeout(timeout)
                .retry(3, false)
                .asJsonObject();
        } catch (WebbException e) {
            errorResponse = e.getResponse();

            if (errorResponse != null) error = new RatingsHTTPError(errorResponse.getStatusCode(), errorResponse.getStatusLine());
            else error = new RatingsHTTPError(112, e.getMessage());

            return new RatingsResult(null, error);
        }

        return new RatingsResult(response.getBody(), null);
    }

    public void setTimeout(Integer value) {
        timeout = value * 1000;
    }

    public void setUser(String name, String mibaId) {
        setUser(name, mibaId, null);
    }

    public void setUser(String name, String mibaId, String email) {
        HashMap<String, String> newUser;

        newUser = new HashMap<String, String>();

        if (name != null) newUser.put("name", name.trim());
        if (mibaId != null) newUser.put("mibaId", mibaId.trim());
        if (email != null) newUser.put("email", email.trim());

        user = newUser;
    }

    public RatingsResult create(int rating) {
        return create(rating, null, null);
    }

    public RatingsResult create(int rating, String description) {
        return create(rating, description, null);
    }

    public RatingsResult create(int rating, String description, String comment) {
        RatingsRequest request;
        RatingsError error;

        validateRating(rating);

        request = getRequest(rating);

        if (description != null) {
            error = validateDescription(description);

            if (error != null) return new RatingsResult(null, error);

            request.description = description.trim();
        }

        if (comment != null) {
            error = validateComment(comment);

            if (error != null) return new RatingsResult(null, error);

            request.comment = comment.trim();
        }

        if (user != null) {
            error = validateUser(user);

            if (error != null) return new RatingsResult(null, error);

            request.user = user;
        }

        return send(request);
    }
}

File: sdks/swift/RatingsSDK/Ratings.swift
//
//  RatingsSDK.swift
//  RatingsSDK
//
//  Created by Rita Zerrizuela on 7/25/17.
//  Copyright © 2017 GCBA. All rights reserved.
//

import Foundation
import SwiftHTTP
import SwifterSwift
import GBDeviceInfo

public class Ratings {
    public init(api url: String, token: String, app: String, platform: String, range: String) {
        let baseUrl = url.trimmed
        
        self.url = baseUrl.lastCharacterAsString == "/" ? baseUrl + "ratings" :  baseUrl + "/ratings"
        self.token = token.trimmed
        self.app = app.trimmed
        self.platform = platform.trimmed
        self.range = range.trimmed
        self.deviceInfo = GBDeviceInfo()
        self.timeout = 3
        
        validateUrl(self.url)
        validateToken(self.token)
        validateKey(self.app, description: "app")
        validateKey(self.platform, description: "platform")
        validateKey(self.range, description: "range")
    }
    
    // MARK: - Private properties
    
    private let url: String
    private let token: String
    private let app: String
    private let platform: String
    private let range: String
    private let deviceInfo: GBDeviceInfo
    
    private var _user: RatingsUser?
    
    // MARK: - Public properties
    
    public var timeout: Double
    public var user: RatingsUser?

    // MARK: - Validations

    private func validateUrl(_ url: String) {
        guard url.isValidUrl else { fatalError("invalid url") }
    }
    
    private func validateToken(_ token: String) {
        guard token.length > 0 else { fatalError("invalid token") }
    }
    
    private func validateKey(_ key: String, description: String) {
        guard key.length == 32 else { fatalError("\(description) is not a valid key") }
    }
    
    private func validateName(_ name: String) -> RatingsError? {
        guard name.length >= 3 else { return RatingsError.nameTooShort("name too short") }
        guard name.length <= 70 else { return RatingsError.nameTooLong("name too long") }
        
        return nil
    }
    
    private func validateEmail(_ email: String) -> RatingsError? {
        guard email.isEmail else { return RatingsError.invalidEmail("invalid email") }
        guard email.length >= 3 else { return RatingsError.emailTooShort("email too short") }
        guard email.length <= 100 else { return RatingsError.emailTooLong("email too long") }
        
        return nil
    }
    
    private func validateMibaId(_ mibaId: String) -> RatingsError? {
        guard mibaId.length == 36 else { return RatingsError.invalidMibaId("invalid mibaId") }
        
        return nil
    }
    
    private func validateRating(_ rating: Int) -> RatingsError? {
        guard rating >= -127 && rating <= 127 else { return RatingsError.invalidRating("invalid rating") }
        
        return nil
    }
    
    private func validateDescription(_ description: String) -> RatingsError? {
        guard description.length >= 3 else { return RatingsError.descriptionTooShort("description too short") }
        guard description.length <= 30 else { return RatingsError.descriptionTooLong("description too long") }
        
        return nil
    }
    
    private func validateComment(_ comment: String) -> RatingsError? {
        guard comment.length >= 3 else { return RatingsError.commentTooShort("comment too short") }
        guard comment.length <= 1000 else { return RatingsError.commentTooLong("comment too long") }
        
        return nil
    }

    // MARK: - Setters
    
    public func validateUser(_ user: RatingsUser) -> RatingsError? {
        guard user.mibaId != nil || user.email != nil else { return RatingsError.missingEmailAndMibaId("user is missing email and mibaId") }
        
        if let error = validateName(user.name) { return error }
        
        if let mibaId = user.mibaId {
            if let error = validateMibaId(mibaId) { return error }
        }
        
        if let email = user.email {
            if let error = validateEmail(email) { return error }
        }
        
        _user = user
        
        return nil
    }
    
    // MARK: - Helpers
    
    private func buildParams() -> [String: Any] {
        let screen: [String: Any] = [
            "width": UIScreen.main.bounds.width,
            "height": UIScreen.main.bounds.height,
            "ppi": deviceInfo.displayInfo.pixelsPerInch
        ]
        
        let device: [String: Any] = [
            "name": deviceInfo.modelString,
            "brand": "Apple",
            "screen": screen
        ]
        
        let result: [String: Any]  = [
            "range": range,
            "app": [
                "key": app,
                "version": Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as! String
            ],
            "platform": [
                "key": platform,
                "version": "\(deviceInfo.osVersion.major).\(deviceInfo.osVersion.minor).\(deviceInfo.osVersion.patch)"
            ],
            "device": device
        ]
        
        return result
    }
    
    private func send(params: [String: Any], retry: Int = 3, callback: @escaping (_ response: Response?, _ error: RatingsError?) -> ()) {
        let headers = [
            "Content-Type": "application/json; charset=UTF-8",
            "Accept": "application/json",
            "Authorization": "Bearer \(token)"
        ]
        
        let backoff = 3 - retry
        
        HTTP.globalRequest { [unowned self] (request: NSMutableURLRequest) in
            request.timeoutInterval = self.timeout
        }
        
        DispatchQueue.global(qos: .userInitiated).asyncAfter(deadline: .now() + .seconds(backoff)) {
            do {
                try HTTP.POST(self.url, parameters: params, headers: headers, requestSerializer: JSONParameterSerializer()).start { response in
                    guard response.error == nil else {
                        if (response.error!.code == 503 || response.error!.code == 504 || response.error!.code >= 520) && retry > 0 {
                            self.send(params: params, retry: retry - 1, callback: callback)
                                
                            return
                        }
                        
                        response.error!.domain == "HTTP" ?
                            callback(nil, RatingsError.http(response.error!.localizedDescription, response.error!.code, response)) :
                            callback(nil, RatingsError.other(response.error!.localizedDescription))
                        
                        return
                    }
                    
                    callback(response, nil)
                    
                    return
                }
            }
            catch let error {
                debugPrint(error.localizedDescription)
                callback(nil, RatingsError.other(error.localizedDescription))
            }
        }
    }
    
    // MARK: - Public API
    
    public func create(rating: Int, description: String? = nil, comment: String? = nil, callback: @escaping (_ response: Response?, _ error: RatingsError?) -> ()) {
        if let error = validateRating(rating) { return callback(nil, error) }
        
        var params: [String: Any] = buildParams()
        
        params["rating"] = rating
        
        if let description = description {
            if let error = validateDescription(description) { callback(nil, error) }
            
            params["description"] = description
        }
        
        if let comment = comment {
            if let error = validateComment(comment) { callback(nil, error) }
            
            params["comment"] = comment
        }
        
        if let user = user {
            if let error = validateUser(user) { callback(nil, error) }
            
            params["user"] = user.toDict()
        }
        
        send(params: params, callback: callback)
    }
}
